TC Custom Service Importer

awisema opened the pull request:
[//]: # 'remove this if PR is for a release-* branch'
# _Please squash merge this PR_

## Description

Importer for custom services. As the exporter has not yet been built, there are probably some adjustments to be made to make it conform to the export.

## Type of change
[//]: # 'remove options that are not relevant'
- [X] New feature (adds functionality)

## Checklist before requesting review
- [X] I have performed a self-review of my code
- [X] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [X] My code includes comments and/or descriptive variable names to help other engineers understand the intent (or not applicable)
- [X] My code follows the style guidelines of this project (rubocop)
- [X] I have updated the documentation (or not applicable)
- [X] If it's not obvious how to test this change, I have provided testing instructions in this PR or the related issue

ttoomey committed:
186922726-tc-eto-importers

ttoomey committed:
Merge branch 'release-103' into 186922726-tc-eto-importers

ttoomey committed:
186922726-tc-eto-importers

ttoomey committed:
186922726-tc-eto-importers

ttoomey committed:
186922726-tc-eto-importers

ttoomey committed:
186922726-tc-eto-importers

ttoomey committed:
186922726-tc-eto-importers

dry up

ttoomey committed:
186922726-tc-eto-importers

adjust spdat / hat data dictionary

ttoomey committed:
186922726-tc-eto-importers

add spdat builder (186786001)

ttoomey committed:
186922726-tc-eto-importers

ttoomey committed:
186922726-tc-eto-importers

ttoomey committed:
186922726-tc-eto-importers

ttoomey committed:
Merge branch 'release-103' into 186922726-tc-eto-importers

ttoomey committed:
186922726-tc-eto-importers

ttoomey committed:
186922726-tc-eto-importers

ttoomey committed:
186922726-tc-eto-importers

ttoomey committed:
186922726-tc-eto-importers

todo or die

ttoomey committed:
186922726-tc-eto-importers

awisema committed:
WIP: Process custom data attached to Client

awisema committed:
WIP: Process custom contact points

awisema committed:
WIP: Process custom contact points

awisema committed:
Merge remote-tracking branch 'origin/186922726-tc-eto-importers' into 186922726-tc-eto-importers

# Conflicts:
#	drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/custom_client_demographics_loader.rb

awisema committed:
WIP: Process custom addresses

awisema committed:
Integrate demographics loader into importer

awisema committed:
Merge branch 'release-104' into 186922726-tc-eto-importers

awisema committed:
WIP: HMIS custom services loader

awisema committed:
WIP: HMIS custom services configuration

ttoomey committed:
aw/service-importer-187021509

small fixes

ttoomey reviewed (approved):
Looks good, thanks! I'll take care of handling the merge

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/custom_services_loader.rb:
```diff
@@ -0,0 +1,525 @@
+
+    private def normalize_contact(contact)
+      contact # TODO
+    end
```
Awesome, would be great to normalize into standard values. From what I saw it is often just the casing and punctuation that differed across services, but it needs a closer look.

Suggestion :

Normalized Service Location options:
* Place not meant for habitation
* Service Setting, Non Residential
* Service Setting, Residential

Normalized Type of Contact options:
* Home
* Phone
* Phone Left Voice Message
* Email
* Office
* In-Person
* Driving Client
* No Show Home Visit
* No Contact/No Show
* Counseling
* Case Management Advocacy
* Other

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/custom_services_loader.rb:
```diff
@@ -0,0 +1,525 @@
+            service_name: service_type,
+            DateCreated: today,
+            DateUpdated: today,
+            FAAmount: nil,
```
Services that collect financial assistance amount should store it in this field, _not_ as a custom data element. I think here you're storing it in data elements with key "FAAmount", which we don't want

gigxz replied:
nevermind, I see they are nested under `service_fields`

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/custom_services_loader.rb:
```diff
@@ -0,0 +1,525 @@
+          service_type: 'Transportation',
+          service_fields: {
+            'Cost' => {
+              key: :FAAmount,
```
See note above, this can be stored directly on the CustomService table

gigxz commented:
thank you Sandy this looks great!!!

awisema committed:
Changes based on ETO export

awisema committed:
Updates and fixes from testing

ttoomey commented:
@gigxz do you think this PR ready to merge? If so I can fix up the conflicts and merge it in

gigxz commented:
@ttoomey I am going to open a PR against this branch with a few suggestions

gigxz commented:
alllright @ttoomey if you can merge in https://github.com/greenriver/hmis-warehouse/pull/4093 and then resolve merge conflicts and merge to `release-107` that would be awesome!

Hopefully I caught everything in that PR. I did an export of the last few months and compared the field names. (A few hadn't been used in the last few months so I needed to run a longer history to look at those: Baby Supplies; Benefits Contacts; Budgeting/FInancial Planning; and Food Box ).

gigxz committed:
Match field names and normalize location (#4093)

* Fix incorrect date types on ce event (#4022)

* TC CAS Readiness (#3991)

* 186994227-log-graphql-api-response-time (#4014)

* 186994227-log-graphql-api-response-time

* 186994227-log-graphql-api-response-time

remove unused yabeda gem

* 186994227-log-graphql-api-response-time

* Synthetic events should not be duplicated (#4023)

* Fix non-functional active record scopes

* Apply suggestions from code review

Comment updates

Co-authored-by: Sandy Wise <33072305+awisema@users.noreply.github.com>

---------

Co-authored-by: Sandy Wise <33072305+awisema@users.noreply.github.com>

* Adjustments for Pathways 2024

* Add pathways upload feed (#4017)

* Add pathways upload feed

* Private def

* Address feedback

* Fix uniqueness constraint violation in HMIS client merge (#4018)

* Fix uniqueness constraint violation in HMIS client merge

* Update drivers/hmis/app/jobs/hmis/merge_clients_job.rb

Co-authored-by: Theron Toomey <ttoomey@greenriver.org>

---------

Co-authored-by: Theron Toomey <ttoomey@greenriver.org>

* Update developer_setup.md

Adjust to colima and traefik

* CE relaunch fixes (#4028)

* Include total night unsheltered

* Fix for default translation

* More adjustments to default translations

* Don't use Yes to represent 1 always

* Don't use Yes to represent 1 always

* Update Core Demographic report to include HIV counts without Indefinite and Impairs. Prevent details from being viewed when user does not have access. (#4027)

* Update Core Demographic report to include HIV counts without Indefinite and Impairs. Prevent details from being viewed when user does not have access.

* update details view to positive check

* Update can_view_details check for better readability

* Clean up disability query

* Update secure files to show when the file will be removed (#4030)

* Tree Menus (#4016)

* WIP: stubs for menu structure

* WIP: more scaffold

* non-recursive menu printing

* Menu system

* Catch nil

* Don't add items to the menu you can't see; move menu into a class so we can create multiple menus

* Open section when in a section; underline on mouse over

* Provide a mechanism to identify active section with regular expressions

* Typo fix; Remove extraneous file; Accessiblitiy improvemnt; Safer regex

* Don't show top-level items if they should not be visible; check HmisEnforcement before accessing path helper

* Scaffold for overrides

* New tests that all fail

* Import override implementation and tests passing

* Break applies? out of apply for future work on exposing overrides in the UI

* Expose overrides on source HMIS pages and project show

* Auto exit bug fix: release unit and notify LINK (#4025)

* Enable sort on project services and households (#4026)

* Enable sort on project services and households

* Use merge instead of arel for household HoH sorting

Co-authored-by: Elliot <elliot@greenriver.org>

* Update all household sorts to use merge instead of arel

* Revert change to warehouse_structure.sql

---------

Co-authored-by: Elliot <elliot@greenriver.org>
Co-authored-by: Gig <gig@greenriver.org>

* Add Signature dynamic field component (#3997)

* Fix for menu not remembering location; update order of reports menu

* Move Favorites to the top of report page

* Make client alerts auditable (#3998)

* Clean up report date range (#4001)

* Use household lookup instead of SHE link

* Fix for re-use of household IDs across projects

* Fix for weird household data

* Use more common language

* Remove extraneous language

* Use client ts to lookup CE events instead of enrollments

* set cohort download as a background job

* include additional params to allow different versions of the cohorts excel export to be created.

* Superset docker-tweak

* String-ify params keys, remove old XLSX format mechanism and digest cache key

* 187020495-time-of-day-input (#4031)

* 187020495-time-of-day-input

* 187020495-time-of-day-input

typo

---------

Co-authored-by: Gig <gig@greenriver.org>

* Add time duration component to dynamic forms (#4038)

* Add time duration component to dynamic forms

* Rename TIME_DURATION to MINUTES_DURATION to be more specific

* Support general input table for forms (#4041)

* Add Unique index to import overrides

* Simplify the presence check

* Run import overrides in a consistent order with the least specific first

* Correctly identify data source id

* Provide a mechanism to force a menu section to be always open

* Update README.md

* Update README.md (#4048)

* Don't show HUD Reports in the menu if you can't get there

* Remove PDDE overrides in favor of import overrides (#4034)

* Updates to the Project Type Reconcilliation report

* WIP; Remove additional overrides for project type and update project cleanup

* Remove references to old override columns

* Missed file

* Fixes for bad fixes

* Adjustments for failing tests

* Fix for recent service history table

* Removing overrides from external api import; no longer necessary

* TodoOrDie for removing unused override columns; typo fixes

* Export should add Move-in dates for enrollments at projects that weren't PH but have been overridden to PH; test to confirm this

* Fix for production missing route

* Security updates (#4049)

* Security updates

* Revert other updates; they were unintended

* Hotfix for hidden range bugs (#4053)

* add excel_export variable for controllers with nav section in cohorts page (#4054)

* 186922726 additional eto imports (#4009)

* create "synthetic form processors" for imported assessments
* diversion importer
* navigation notes importer
* cleanup columns
* add element ids to hat col config when labels don't match
* clean up row value by id vs by label
* includes EVH App importer, built this but sounds like we won't use it. Leaving it there for now in case we discover a need for it.

* Build updates  (devops issue 685) (#4029)

* Updates/improvements while working on CAS

* Have both CA bundles available

WIP: follow up with ansible search

* Don't need pre-cache-version anymore

* typos and doc updates

* Remove defunct files

* add back the old CA

* Secrity update for actionpack

* Rack security update; ignore CVE that isn't relevant and doesn't have a fix

---------

Co-authored-by: Elliot <elliot@greenriver.org>

* Custom demographics import (#4024)

* Import for custom ETO client demographics.
* use roo streaming mode to avoid memory issues with large sheets
* fix for extra characters in custom assessment enrollment ids

---------

Co-authored-by: Theron Toomey <ttoomey@greenriver.org>

* Phase 2 assessment importers (#4043)

* WIP: Critical Documents Case Management Loader

* WIP: Case Management Assessment Loader

* WIP: MHMR Case Management Note Loader

* WIP: MHMR Non-Billable Note Loader

* WIP: MHMR Rehabilitation Note Loader

* DRY up generation of MHMR service code rows

---------

Co-authored-by: Theron Toomey <ttoomey@greenriver.org>

* Make move-in date within report range for ANSD (#4056)

* Generate a separate claim for each QA (#3917)

* Generate a separate claim for each QA

* QA Modifier Updates
Add collateral contact tp T2038
Add telehealth to T2038>U5

(cherry picked from commit 174c64016acfe97cb2c033e1e9f8aa919ede8a26)

* Only code calls/video with the client as telehealth (#4040)

* Dynamic POS (#4042)

* Generate a separate claim for each QA

* QA Modifier Updates
Add collateral contact tp T2038
Add telehealth to T2038>U5

(cherry picked from commit 174c64016acfe97cb2c033e1e9f8aa919ede8a26)

* Push claim POS (place of service) out to QAs

* Adjust telehealth POS codes

* Add Auto-Entry functionality for non-HUD projects (#4044)

* active_cohort column breaking on download when cohort names does not exist (#4061)

* active_cohort column breaking on download when cohort names does not exist

* Add cohort_names variable for cohort excel export

* 186922726-add-contributions-table-back-to-ua-import (#4059)

* TC HAT pregnancy status - always family (#4062)

* Mark all clients who indicated they are pregnant as a family

* Mark clients as part of a family if they are currently pregnant

* Add checks for enrollments without clients to the SPM report (#4060)

* Add NC specific SCSS file (#4069)

* Make client alert expiry minimum tomorrow (#4068)

* Make client alert expiry minimum tomorrow

* Use today with offset instead of tomorrow

---------

Co-authored-by: Gig <gig@greenriver.org>

* Update Auto-Exit code to consolidate with Project Configs (#4045)

* Add ProjectConfig with single-table inheritance

* Update project config models and tests

* Add behavior to save non-wip enrollment

* Update Auto-Exit code to consolidate with Project Configs

* Remove errant todo

* Update todo with reminder to remove old table

* Rollback changes to structure.sql

* Only check auto-enter config for new enrollments

Co-authored-by: Gig <gig@greenriver.org>

* Save a synthetic intake assessment on auto-enter

* Remove option for default project config

* Use scope instead of class method for configs_for_project

Co-authored-by: Gig <gig@greenriver.org>

* Update usage of configs_for_project scope

* Add scope for active project configs

* Fix failing test

* Use bang operator for creating auto exit

Co-authored-by: Gig <gig@greenriver.org>

* Update auto exit migration to run in transaction

* Update validation for project configs

* Respond to PR feedback

* Fix failing tests

* Fix failing test

* Fix tests and make options more flexible for future usage

* Fix failing test

---------

Co-authored-by: Gig <gig@greenriver.org>

* Clean up old AutoExitConfig table (#4046)

* Add Cohort Columns for Sheltered and Unsheltered Days Homeless (#4066)

* Add Cohort Columns for Sheltered and Unsheltered Days Homeless

* include sheltered and unsheltered days homeless for cohort clients in the refresh_time_dependant_client_data calculations

* structure update

---------

Co-authored-by: Elliot Anders <elliot@greenriver.org>

* Role Manager (#4067)

* WIP: stubs for menu structure

* WIP: more scaffold

* non-recursive menu printing

* Menu system

* Catch nil

* Don't add items to the menu you can't see; move menu into a class so we can create multiple menus

* Open section when in a section; underline on mouse over

* Provide a mechanism to identify active section with regular expressions

* WIP; scaffold for new role manager

* Typo fix; Remove extraneous file; Accessiblitiy improvemnt; Safer regex

* WIP; additional role bits

* Don't show top-level items if they should not be visible; check HmisEnforcement before accessing path helper

* WIP; more work on roles page

* WIP

* WIP

* Stimulus controller and synchronized sections

* Toggle columns and sections

* Role permissions save; need saving UI

* Save with toasts

* Save button with changed status

* Remaining role functionality

* Better programmatic colors

* Port role changes to HMIS admin

* Fix for title

* Typo fix

* Move headers into column

* Design adjustments from review

* Wording changes

* Remove 2022-2024 transformation from MPER import (#4065)

* Fix bug with CE Assessment and Event deletion (#4064)

* Fix bug with CE Assessment and Event deletion

* Fix typo in test

* Respond to PR feedback

* Fix typo

---------

Co-authored-by: Gig <gig@greenriver.org>

* Make AutoExit and AutoEnter generic into ProjectConfig (#4055)

* Fixes incorrect timing of scheduled tasks (cron) (#4074)

* Fixes incorrect timing of scheduled tasks (cron)

* Run cron installer in the rails context

* Typo fix

* Add new option and functionality to MigrateAssessmentsJob (#4076)

* Always make HUD User optional (#4082)

* Require that client alert expiration date be in the future (#4073)

Co-authored-by: Gig <gig@greenriver.org>

* Remove old AutoExitConfig model (#4075)

* Remove old AutoExitConfig model

* Fix incorrect reference in CreateProjectConfig

* Remove import overrides that would have done nothing in previous version (#4083)

* Update categories and descriptions for HMIS permissions (#4084)

* Update HMIS link style, only show if user has access (#4085)

* Role Manager updates (#4078)

* Include sub-category title when searching roles; hide empty sub-categories when searching

* Add a note that there are changes pending inline

---------

Co-authored-by: Elliot Anders <elliot@greenriver.org>

* Superset roles [devops 702] (#4071)

* Possible superset role config

* Look for either version of green river email

* Superset config moves to superset-sync repo

* roles added to payload

* Roles aren't stable or recommended for this

* superset roles

* Improves documentation

---------

Co-authored-by: Elliot Anders <elliot@greenriver.org>

* Add edit links to role manager

* Fix for ANSD publishing after JS changes and updates to Perfromance Measurement publlish summary feature (#4089)

* Update schema files for tests

* HRSN QAs from individual homelessness updates (#3994)

* Generate a separate claim for each QA
* Per-patient, per-day housing statuses
* Collect per-day housing status information from data sources
* Use recorded homeless status in claims
* Add daily SDoH QAs based on homeless status
* Simple test to for creating daily homeless QA
* Limit daily HRSN QAs
* Test regressions.

* Label custom assessments in audit events (#4063)

* Label custom assessments in audit events

* Add test to assessment name audit

* Avoid n+1 issue when loading custom assessment title

Co-authored-by: Theron Toomey <ttoomey@greenriver.org>

* Use changed_record instead of object for ar assoc

* Use customassesment instead of changed_record

---------

Co-authored-by: Theron Toomey <ttoomey@greenriver.org>
Co-authored-by: Gig <gig@greenriver.org>

* Match field names and normalize location

* SPM updates (#4079)

* WIP: Measure 1 datalab fixes

* WIP: Measure 1 datalab fixes

* Measure 3.2 passes; others uncommented but failing

* WIP: Measure 1a datalab fixes

* Slightly better M1b counts; adjustments to who is included

* Adjust manual spec

* Include time prior to lookback if appropriate

---------

Co-authored-by: Sandy Wise <sandy@greenriver.org>

* Disabling migration for non-existent class

* Don't round until display time

* gig/proposed-service-importer-fix

handle newline in question col

* gig/proposed-service-importer-fix

---------

Co-authored-by: Sandy Wise <33072305+awisema@users.noreply.github.com>
Co-authored-by: Theron Toomey <ttoomey@greenriver.org>
Co-authored-by: Elliot <elliot@greenriver.org>
Co-authored-by: Dave G <149399758+dtgreiner@users.noreply.github.com>
Co-authored-by: martha <medwards@greenriver.org>
Co-authored-by: Dave G <dgreiner@greenriver.org>
Co-authored-by: Todd <tdblackman@gmail.com>
Co-authored-by: Sandy Wise <sandy@greenriver.org>

ttoomey committed:
Merge branch 'release-108' into aw/service-importer-187021509

ttoomey committed:
service-importer-187021509

ttoomey commented on drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/custom_services_loader.rb:
```diff
@@ -0,0 +1,578 @@
+          end
+
+          row_field_value = row_enrollment_id(row)
+          next if row_field_value.blank? # enrollment id is blank
```
@gigxz something to flag: I exported everything from 2023 to present (selecting just the touchpoints used in this importer) and it looks like about 50% of the records had a blank enrollment_id

gigxz replied:
@ttoomey thank you - that is good to know - I will flag to them

ttoomey commented:
Fixes:
* retarget this to release-108
* use only the first line from Question
* create missing service types in development mode
* don't invoke cleaner if the answer value is empty
* clean curly braces out of enrollment_id
* fix service type name s/1A-SA Dinner/1-SA Dinner/

ttoomey commented:
@gigxz I made a few additional changes, could use a quick review before merge. Thanks!

ttoomey committed:
service-importer-187021509

fix logging bug

gigxz reviewed (approved):
looks good to me! thanks

ttoomey committed:
aw/service-importer-187021509

weird timing issue in test failure?
